### DISCUSSION TITLE
keychain to keep reputation

### DISCUSSION THREAD

**Discussion by grunch** [MostroP2P/mostro#362] [Ideas]
As we implemented nip59 to Mostro which will allows more privacy to users, I'm thinking how to keep privacy and still allow users to be rated and share their reputation.

Here's an idea that we can discuss here:

Clients will always use a new ephemeral key for each new order, the user who wants to maintain reputation can send an event to Mostro signed with the last ephemeral key used to update their rating.

Example:

* Alice uses the automatically generated key abc to create a sale order
* Bob takes the order, they finish the operation
* Each one rated their counterparty's ephemeral pubkey

later:

* Alice wants to make another sale, the client automatically generates the key xyz
* Alice's client automatically sends a message inside a gift wrap event to Mostro signed with the key xyz and the internal content signed with the key abc, this way Mostro knows that the user of xyz is also the user of abc
* Alice creates the order with the key xyz and Mostro shows the reputation obtained with abc
* Carol takes the order
* They finish the operation
* Each one rated the counterparty's ephemeral pubkey (xyz in case of Alice)

Mostro only need to save the last Alice's pubkeys to keep the chain

**Comment by bilthon** [comment]
If the ultimate objective is to selectively reveal one's identity to mostro (and only mostro) in order to build a reputation, it seems to me like an overkill to do this chain. A similar result would be obtained if the user were just to use always the same key at the seal level.

**Reply by grunch** [reply]
> If the ultimate objective is to selectively reveal one's identity to mostro (and only mostro) in order to build a reputation, it seems to me like an overkill to do this chain.

Not even Mostro knows the identity of the user, Mostro only knows that all ephemeral keys are own by a user but everytime the user is using a new ephemeral, there is no a common key to link any user.

> A similar result would be obtained if the user were just to use always the same key at the seal level.

If the user uses the same key at seal level all counterparts will know the same identity

**Reply by bilthon** [reply]
> If the user uses the same key at seal level all counterparts will know the same identity

I didn't mean to use the same key at the seal level with all counterparts. I thought this was limited to **client <-> mostro** communications.

> Not even Mostro knows the identity of the user, Mostro only knows that all ephemeral keys are own by a user but everytime the user is using a new ephemeral, there is no a common key to link any user.

Well this is a good point. I guess we could do it, I'm just thinking (like @Catrya hinted at below) about the implications of this for portability. Could clients maybe use something like [NIP-06](https://github.com/nostr-protocol/nips/blob/master/06.md) to manage the keys and make sure they're all taken when the user moves clients?

**Reply by grunch** [reply]
> I didn't mean to use the same key at the seal level with all counterparts. I thought this was limited to **client <-> mostro** communications.

But if we use the same key at the seal level, how Mostro will put them to communicate to each other without revealing that pubkey? the other key (the one used to wrap the seal) is ephemeral
 

> 
> Well this is a good point. I guess we could do it, I'm just thinking (like @Catrya hinted at below) about the implications of this for portability. Could clients maybe use something like [NIP-06](https://github.com/nostr-protocol/nips/blob/master/06.md) to manage the keys and make sure they're all taken when the user moves clients?

I guess that is possible too

**Reply by bilthon** [reply]
> But if we use the same key at the seal level, how Mostro will put them to communicate to each other without revealing that pubkey? the other key (the one used to wrap the seal) is ephemeral

The client could add a "trade key" when creating an offer, and the taker could provide this key to Mostro when taking an order. This way, the communication between the two parties would happen via this pair of dedicated ephemeral "trade keys."

While chaining keys is an interesting concept, at the end of the day Mostro will know that a chain of keys and all related trades are from a single entity, so we don't gain much from it. Moreover, implementing such a mechanism would introduce additional complexity on the Mostro side, so it might be more efficient to use a single key to communicate with Mostro

Additionally note that when I suggest using the same key to interact with Mostro, I don't mean using the same key the user uses for social media or other activities. Ideally, a separate key should be generated for Mostro-specific interactions. We could use NIP-06 to derive a dedicated key for Mostro, providing some separation between the user's public identity and their trades.

**Reply by grunch** [reply]
> The client could add a "trade key" when creating an offer, and the taker could provide this key to Mostro when taking an order. This way, the communication between the two parties would happen via this pair of dedicated ephemeral "trade keys."

this is what I proposed in this discussion
 
> While chaining keys is an interesting concept, at the end of the day Mostro will know that a chain of keys and all related trades are from a single entity, so we don't gain much from it. Moreover, implementing such a mechanism would introduce additional complexity on the Mostro side, so it might be more efficient to use a single key to communicate with Mostro

we are following nip59, can you design how those events look like and show it here?  take any example from http://mostro.network/messages and adapt it to your proposal, that way I can see more clear what are your thoughts

**Reply by bilthon** [reply]
Ok, lets take a step back. There are 2 kinds of communications here:

1- Client-to-Client
2- Client-to-Mostro

The first one I think is straightforward. No need to maintain a long term identity here, we just need a per-trade one. So ephemeral "trade keys" are all we need.

It's in the second one where some sort of persistent identity might be desirable if the user choses to maintain reputation. Honestly, this could be optional, and if the user doesn't care about reputation (or value their privacy more) they should be allowed to opt out and just use ephemeral keys here as well. So same treatment as with option 1.

It's only when a persistent reputation is desired that Mostro (and only Mostro) needs to track of an entity across multiple trades. The chain of keys you proposed is one way o achieving this. But considering Mostro (and only Mostro) will be able to group all those keys under a single entity anyway, why bother creating the chain? it would be simpler to use a Mostro-specific key, unrelated to social media, for Client-to-Mostro communications. Again, Mostro wouldn’t know the user's identity, as this is a Mostro-specific key. So it's as if we collapsed all the keys in the chain into a single one.

### Details
Since you asked, here are the changes I think would be necessary. The create order message carries a new field called `trade_key`, like this:

```json
{
  "order": {
    "version": 1,
    "action": "new-order",
    "content": {
      "order": {
        "kind": "sell",
        "status": "pending",
        "amount": 0,
        "fiat_code": "VES",
        "min_amount": null,
        "max_amount": null,
        "fiat_amount": 100,
        "payment_method": "face to face",
        "premium": 1,
        "trade_key": "0e486495ee6fe36e23f392f604b9c748be00f9a164473552e1d5c60b61e1ec13"
        "created_at": 0
      }
    }
  }
}
```

Similarly, when a user takes an order, the message would look like:

```json
{
  "order": {
    "version": 1,
    "id": "ede61c96-4c13-4519-bf3a-dcf7f1e9d842",
    "action": "take-sell",
    "content": {
        "trade_key":"440c4bb06013b6b52208092ad1df9e40fcd2ccc98cf713bbe45133c2dba25dd2"
    }
  }
}
```

The `trade_key` is an ephemeral per-trade key that clients communicate to Mostro, indicating they're willing to use it for P2P communications. Mostro just stores them and uses them to connect the parties after the hold invoice is secured.

### NIP-06

To elaborate on how and why these keys aren't related to social media activity, let's consider how NIP-06 could be used here. This spec essentially proposes a derivation path for Nostr applications.

The registered BIP-44 purpose number is `1237` so the derivation path looks like this:

```
m/44'/1237'/<account>'/0/0
```

If we assume the `account` zero would be used for social media activity, we could just select a specific index at the account level for Client-to-Mostro communications. I propose `129484`, which is the unicode number for the 🧌 emoji. So, the derivation path for this Client-to-Mostro key would be:

```
m/44'/1237'/129484'/0/0
```

### Extra
This could even be extended to include ephemeral keys under the arrangement. Ephemeral keys could be derived from the mnemonic like this:

```
m/44'/1237'/129484'/0/1
m/44'/1237'/129484'/0/2
m/44'/1237'/129484'/0/3
...
```

This improves portability, as all the user would need to enter when moving to a new client is the mnemonic and the index of the last ephemeral key that was used. Mostro could even help with that keeping track of this info and allowing the new client to recover this  ephemeral key index in the new client via some sort of API.

**Reply by grunch** [reply]
Thanks for this detailed post, this is a good idea, we had this before on our first protocol version, the problem is we can't allow users to add a `trade_key` just like that, if they don't give a proof of ownership we are opening an attacking vector on identity theft, I mean that everyone would be able to use other users pubkeys, we solve this with the signature on the wrapper and seal events, we can do the same and add a signature field on the mostro `message` and sign the entire event without the signature field, but I don't know if we are adding too much complexity, what do you think?

**Reply by bilthon** [reply]
But wouldn't this be a somewhat limited attempt at identity theft? The `trade_key` must actually be used to produce signatures for the P2P communication. What would the attacker gain from providing a public key they can't use to generate valid signatures? The P2P communication wouldn’t happen, and the trade would likely be undone quickly, as the counterparty would likely file a dispute after a while.

At best, this could function as a sort of DoS attack. However, since keys and signatures are cheap to produce, having a chain of keys linked by signatures wouldn’t necessarily prevent this either. In fact, an attacker could achieve a similar outcome by simply not engaging with the counterparty at all.

**Reply by grunch** [reply]
I think I misunderstood, I get that the trade_key is the ephemeral key, even if we are use this key to make them to talk to each other not validating the owner of the key is a bad idea, scammers are really creative and they will come with triangulation attacks where they make a user talk with a third one, I would not risk the security of users accepting a key that I am not validating.

Still I don't know how the user send to Mostro the real pubkey in your proposal.

If your proposal is to include the real user pubkey in the seal's pubkey, we need to validate the trade key with a signature.

**Reply by bilthon** [reply]
> scammers are really creative and they will come with triangulation attacks where they make a user talk with a third one, I would not risk the security of users accepting a key that I am not validating.

In a triangulation scheme the attacker wants to keep the third party out of the loop, not put them in contact with the seller, so I don't think that concern applies. But fine, if you want to tighten things up adding an extra `trade_key_signature` field would be simple to implement.

> Still I don't know how the user send to Mostro the real pubkey in your proposal.

What do you mean by "real pubkey"? the `trade_key` is the real one. The one they're going to use for P2P communications. They can start sending gift-wrapped messages to each other this way after Mostro has delivered trade keys to each other.

**Reply by grunch** [reply]
> In a triangulation scheme the attacker wants to keep the third party out of the loop, not put them in contact with the seller, so I don't think that concern applies. But fine, if you want to tighten things up adding an extra `trade_key_signature` field would be simple to implement.

It is completely irrelevant if neither you nor I can imagine a potential attack right now, I'm not going to leave any loose ends on this.

> What do you mean by "real pubkey"? the trade_key is the real one. The one they're going to use for P2P communications. They can start sending gift-wrapped messages to each other this way after Mostro has delivered trade keys to each other.

After users starts a dispute the client should make them chat sending nostr events with kind 14 and the content encrypted with nip44 algorithm, this messages leaks metadata exactly the same like nip04, but with nip44 we have the option of  allow users to export their conversation key and voluntarily gave it to an admin if they need, having the opportunity of read the messages from both parties the admin can easily solve the dispute.

But if a user uses the same key for all trades the conversation key would be the same if the trade each other again, so the admin could see older messages, that's why we should force users to use a ephemeral key for each trade.

As this proposal is about chain ephemeral keys (trade_key) used on every order to allow to the user to be rated without leaking one potential personal key, and your proposal to improve this is to use the same key in all orders to identify the user only against mostro, that would be what I call `the real pubkey`, I guess that we can show this pubkey on nostr to link with the rating but this key shouldn't be used to talk to users to avoid surveillance.

**Reply by bilthon** [reply]
> It is completely irrelevant if neither you nor I can imagine a potential attack right now, I'm not going to leave any loose ends on this.

Cool, I'm open to adding a signature field.

> But if a user uses the same key for all trades the conversation key would be the same if the trade each other again, so the admin could see older messages, that's why we should force users to use a ephemeral key for each trade.

Yes, that's why I called them trade keys. You get to give Mostro a new one for each trade. If you happen to trade with the same user again you both will have new keys.

> As this proposal is about chain ephemeral keys (trade_key) used on every order to allow to the user to be rated without leaking one potential personal key, and your proposal to improve this is to use the same key in all orders to identify the user only against mostro, that would be what I call the real pubkey, I guess that we can show this pubkey on nostr to link with the rating but this key shouldn't be used to talk to users to avoid surveillance.

Ah ok, so you mean the Client-to-Mostro key. This key should only not be rotated in cases where the client wants to persist an identity in order to gain a rating. But this should be internal as well. No need to publish what the key is, just the reputation with the orders this entity creates.

**Reply by Catrya** [reply]
> But this should be internal as well. No need to publish what the key is, just the reputation with the orders this entity creates.

¿Pero si no se publica esa key, y solo la guarda un Mostrod, cuando hayan varios Mostros, si el usuario cambia de Mostro, no estaría sin reputación en ese otro? creo que se debería mantener una reputación global sin importar el mostro que uses

**Reply by grunch** [reply]
Exactly, the intention of publishing the reputation in nostr is to allow interoperability between Mostros

**Reply by bilthon** [reply]
My approach was based on an equivalence principle—thinking that, from Mostro's perspective, it’s essentially the same whether clients use a chain of keys to represent their identity or a single key consistently. However, that equivalence breaks down if you want to allow clients to port their reputation across different Mostro instances. In that case, you do gain some privacy benefits by constantly switching keys.

The idea of users porting their reputation across Mostro instances is a use case I hadn’t considered. But I think you should be cautious with this approach, as it could quickly become an attack vector. Anyone can run a Mostro instance, and since keys and signatures are cheap to produce, allowing users to transfer their reputation from any other Mostro instance could open the door for people to game the reputation system.

I don’t want to extend this conversation too much. If you want to go with the chain of keys approach, I think it is indeed a more flexible solution and allows for these kinds of features to be introduced. But if the goal is to restrict reputation to a single Mostro instance, then it’s basically equivalent to always using a single key. In that case, a KISS solution is always preferred.

**Reply by Catrya** [reply]
> Anyone can run a Mostro instance, and since keys and signatures are cheap to produce, allowing users to transfer their reputation from any other Mostro instance could open the door for people to game the reputation system.

I hadn't thought about that case, but you're right. We would have to think about whether it can be avoided and at the same time maintain a global reputation, or not.

**Reply by grunch** [reply]
> I hadn't thought about that case, but you're right. We would have to think about whether it can be avoided and at the same time maintain a global reputation, or not.

I had this on mind since the beginning, one of the final steps is to implement a reputation system for users to rate Mostros, Users cannot blindly trust from one day to another in a Mostro, trust must be building each

**Reply by Catrya** [reply]
> I had this on mind since the beginning, one of the final steps is to implement a reputation system for users to rate Mostros, Users cannot blindly trust from one day to another in a Mostro, trust must be building each

And what is the idea? That Mostros with more reputation contribute more to a user's reputation than less used Mostros?
Or that in a user's reputation show something like a percentage of how much he accumulated in each monster he used? For example: 4.5 stars, from there 75% he accumulated in Monster X, 20% in Monster Y, and 5% in Monster Z?
Or something else?

**Comment by Catrya** [comment]
If the user wants to change clients but keep their reputation, what would it be like? 
Would the client 1 provide the last private key with reputation and when entering the new client, the client 2 would sign the content of the message sent to mostro with that key and the signed message with the new key generated? Or is this not possible?

**Reply by grunch** [reply]
that's the main idea, to move from clients you only need to export the last private key used and sign with that one from the new client

**Comment by grunch** [comment]
closing this dicussion in favor of https://github.com/MostroP2P/mostro/discussions/396
